lib/ex_unit/lib/ex_unit/runner.ex

```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
     parent_pid = self()
     timeout = get_timeout(config, %{})
 
-    {module_pid, module_ref} =
-      spawn_monitor(fn ->
-        ExUnit.OnExitHandler.register(self())
-
-        case exec_module_setup(test_module) do
-          {:ok, test_module, context} ->
-            finished_tests = Enum.map(tests, &run_test(config, &1, context))
-            send(parent_pid, {self(), :module_finished, test_module, [], finished_tests})
-
-          {:error, test_module} ->
-            failed_tests = Enum.map(tests, &%{&1 | state: {:invalid, test_module}})
-            send(parent_pid, {self(), :module_finished, test_module, failed_tests, []})
-        end
-
-        exit(:shutdown)
-      end)
+    {module_pid, module_ref} = spawn_module_monitor(config, test_module, parent_pid, tests)
 
-    {test_module, pending, finished_tests} =
+    {test_module, test_results} =
       receive do
-        {^module_pid, :module_finished, test_module, failed_tests, finished_tests} ->
-          Process.demonitor(module_ref, [:flush])
-          {test_module, failed_tests, finished_tests}
+        {^module_pid, :module_finished, test_module, tests} ->
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            Process.demonitor(module_ref, [:flush])
+            {test_module, %{pending: [], finished: []}}
+          else
+            process_failure(config, test_module)
+            Process.demonitor(module_ref, [:flush])
+            {test_module, tests}
+          end
+
+        {^module_pid, :max_failures_reached, test_module, tests} ->
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            Process.demonitor(module_ref, [:flush])
+            {test_module, %{pending: [], finished: []}}
+          else
+            Process.demonitor(module_ref, [:flush])
+            {test_module, tests}
+          end
 
         {:DOWN, ^module_ref, :process, ^module_pid, error} ->
-          test_module = %{test_module | state: failed({:EXIT, module_pid}, error, [])}
-          {test_module, [], []}
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            {test_module, %{pending: [], finished: []}}
+          else
+            test_module = %{test_module | state: failed({:EXIT, module_pid}, error, [])}
+            process_failure(config, test_module)
+            {test_module, %{pending: [], finished: []}}
+          end
       end
 
-    {exec_on_exit(test_module, module_pid, timeout), pending, finished_tests}
+    test_module = exec_on_exit(test_module, module_pid, timeout)
+    {test_module, test_results}
+  end
+
+  defp spawn_module_monitor(config, test_module, parent_pid, tests) do
+    spawn_monitor(fn ->
+      module_pid = self()
+      register(module_pid)
```

Comment:
We are missing information here: "on exit".

Reply:
In addition, I prefer `self()` because it always means "the current process"

Reply:
@fertapric what do you exactly mean by missing on exit information?

Reply:
The previous code was `ExUnit.OnExitHandler.register(pid)` and now is `register(pid)`. Before it was clear that you were registering this process for something related to “on exit” callbacks.

Reply:
True. It was a left over from when we were also registering the PIDs in the runnerstats